driver/src/orderbook/streamed/updating_orderbook.rs

```diff
@@ -0,0 +1,75 @@
+use super::*;
+use crate::contracts::stablex_contract::batch_exchange;
+use anyhow::{anyhow, bail, Result};
+use block_timestamp::BlockTimestamp;
+use ethcontract::{contract::Event, errors::ExecutionError};
+use futures::stream::{BoxStream, StreamExt as _};
+use orderbook::Orderbook;
+use std::future::Future;
+use std::sync::{mpsc, Arc, RwLock};
+
+// TODO:
+// `pub struct UpdatingOrderbook` that also owns an `Arc<RwLock<Orderbook>>` and creates a
+// background thread running `UpdatingOrderbookThread::update_with_events`.
+// This way `UpdatingOrderbookThread` is independently testable.
+
+/// Update the orderbook with events from the stream forever or until exit_indicator is dropped.
+///
+/// Returns Ok when exit_indicator is dropped.
+/// Returns Err if the stream ends.
+async fn update_with_events_forever(
+    orderbook: Arc<RwLock<Orderbook>>,
+    mut block_timestamp: impl BlockTimestamp,
+    exit_indicator: mpsc::Receiver<()>,
+    past_events: impl Future<Output = Result<Vec<Event<batch_exchange::Event>>, ExecutionError>>,
+    mut stream: BoxStream<'static, Result<Event<batch_exchange::Event>, ExecutionError>>,
+) -> Result<()> {
+    let should_exit = || {
+        matches!(
+            exit_indicator.try_recv(),
+            Err(mpsc::TryRecvError::Disconnected)
+        )
+    };
+
+    for event in past_events.await? {
+        handle_event(&orderbook, &mut block_timestamp, event).await?;
+    }
+
+    while !should_exit() {
```

Comment:
This might cause issues if the `stream` never produces events then you will never exit. Alternatively you can use `select!` to choose what comes first between the `exit_indicator` and the `stream`'s next event. 

I would also suggest `select!`ing on the `past_events` as well in case an exit request comes in before the `past_events` query is done.

Reply:
This is an example of select in action:
https://github.com/gnosis/dex-services/blob/1b2beca83f5dc7aeb1205249d74713dd753175bb/driver/src/orderbook/retrying_retrieval.rs#L158-L164

Reply:
Your async tips are very helpful. I forgot about select.